Examples/Jax/jax_example.py

```diff
@@ -1,41 +1,26 @@
 import netket as nk
 import numpy as np
 import jax
-from jax.experimental import stax
-from jax.experimental.stax import Dense, Tanh
+import cProfile
+from jax.config import config
 
+# config.update("jax_log_compiles", 1)
 # 1D Lattice
 L = 20
 g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.PySpin(s=0.5, graph=g)
 
 ha = nk.operator.Ising(h=1.0, hilbert=hi)
 
+alpha = 1
+ma = nk.machine.JaxRbm(hi, alpha, dtype=complex)
```

Comment:
Just to get a better feeling of the performance differences we are still talking about here.

If I replace `JaxRbm` by `RbmSpin` and `JaxMetropolisLocal` by `MetropolisLocal`, I get the following:
```
# For RbmSpin
100%|█████████████████████████████████████| 1/1 [00:02<00:00,  2.50s/it, Energy=-19.9-0.0j ± 0.14 [var=2.0e+01, R_hat=0.9976]]
100%|██████████████████████████████| 300/300 [00:18<00:00, 16.58it/s, Energy=-23.39+0.00j ± 0.062 [var=3.9e+00, R_hat=1.0051]]
# Same with JaxRbm and JAX sampler
100%|█████████████████████████████████████| 1/1 [00:05<00:00,  5.19s/it, Energy=-19.9+0.0j ± 0.14 [var=2.1e+01, R_hat=0.9993]]
100%|███████████████████████████| 300/300 [01:21<00:00,  3.67it/s, Energy=-25.482-0.003j ± 0.0099 [var=9.8e-02, R_hat=1.0006]]
```

So, JAX here has a speed of about  `3.7 it/s` whereas `RbmSpin` has `16.6 it/s` which is 4.5 times faster. Is this in line with your benchmarks for this system size (`N=20` is the example)?

Reply:
No, it is performing a bit better on my laptop... Are you on conda in both cases?

The complex case is slower (as I was also writing) but the dtype=float case on my laptop has basically the same speed on this benchmark. If you instead benchmark **sampling times only** the dtype=float case is actually maybe a factor of 2 faster than the old implementation and doesn't have small n_chains overhead. But it's interesting to see if these considerations extend beyond my laptop... ehe

Reply:
My 2 cents on this. I've run the same example on our cluster (with Intel Skylake, realistic conditions for VMC calculations). I obtain this:
JAX : about `3.4 it/s` (`complex`), about `6.3 it/s` (`float`)
RBM : about `24 it/s` (for both `complex` and `float`, strangely, maybe I missed something)
All this only on one thread.

I wanted to try on the GPU partition, but couldn't get it to work (not an expert though)

Reply:
Thanks @fabienalet , were you in a conda environment? (And no, I wouldn't expect the gpu backend to work already, but it should work for the sampling only though...)

Reply:
No conda, plain python (3.6.3, as installed and compiled optimally on this cluster).
For GPU, I didn't try very much ...

Reply:
It would be great if you guys could benchmark sampling only though, because VMC has overhead from many other things that are now not optimized for non-numpy types! Basically just do something like 

```python
import netket as nk
import numpy as np
import time


# 1D Lattice
L = 20
g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)

# Hilbert space of spins on the graph
hi = nk.hilbert.PySpin(s=0.5, graph=g)

# Ising spin hamiltonian
ha = nk.operator.Ising(h=1.0, hilbert=hi)

# RBM Spin Machine
alpha = 1
dtype = float
ma = nk.machine.RbmSpin(alpha=alpha, hilbert=hi, dtype=dtype)
ma.init_random_parameters(seed=1234, sigma=0.01)


j_ma = nk.machine.JaxRbm(alpha=alpha, hilbert=hi, dtype=dtype)
j_ma.init_random_parameters(seed=1234, sigma=0.01)

# Metropolis Local Sampling
n_chains = 8


j_sa = nk.sampler.JaxMetropolisLocal(machine=j_ma, n_chains=n_chains)
sa = nk.sampler.MetropolisLocal(machine=ma, n_chains=n_chains)

n_samples = 1000 // n_chains


def bench(n_times, sampler):
    for i in range(n_times):
        ma.init_random_parameters(seed=1234, sigma=0.01)
        samples = sampler.generate_samples(n_samples)
    return samples


def j_bench(n_times, sampler):
    for i in range(n_times):
        j_ma.init_random_parameters(seed=1234, sigma=0.01)
        samples = sampler.generate_samples(n_samples)
        samples.block_until_ready()
    return samples


samples = j_bench(1, j_sa)
samples.block_until_ready()
t0 = time.time_ns()
samples = j_bench(300, j_sa)
print("Jax sampler (dtype " + str(dtype) + ")")
tf = time.time_ns()
print("time (s) ", (tf - t0) / 1.0e9)


samples = bench(1, sa)
t0 = time.time_ns()
samples = bench(300, sa)
print("Numpy sampler (dtype " + str(dtype) + ")")
tf = time.time_ns()
print("time (s) ", (tf - t0) / 1.0e9)
```
For n_chains=8 I get 
```
Jax sampler (dtype <class 'float'>)
time (s)  6.190638
Numpy sampler (dtype <class 'float'>)
time (s)  11.375966
```

and 

```
Jax sampler (dtype <class 'complex'>)
time (s)  13.379895
Numpy sampler (dtype <class 'complex'>)
time (s)  13.660763
```

for n_chains=2 
the complex case is 

```
Jax sampler (dtype <class 'complex'>)
time (s)  14.477795
Numpy sampler (dtype <class 'complex'>)
time (s)  36.393191

```

Reply:
On the same cluster, 1 thread
nchains = 8
```Jax sampler (dtype <class 'float'>)
time (s)  14.394850015640259
Numpy sampler (dtype <class 'float'>)
time (s)  14.192219495773315
Jax sampler (dtype <class 'complex'>)
time (s)  26.750790119171143
Numpy sampler (dtype <class 'complex'>)
time (s)  18.07265877723694
```
nchains = 2 
```
Jax sampler (dtype <class 'complex'>)
time (s)  28.338298797607422
Numpy sampler (dtype <class 'complex'>)
time (s)  43.01277542114258
```

Reply:
OK, better!! @fabienalet I think that most likely the residual difference is due to jax maybe not using the best lapack if not in a conda environment, but I might be wrong... 
In any case this seems quite good already, especially because numpy objects are not automatic-differentiable whereas jax stuff is!

Reply:
On my computer:

Jax sampler (dtype <class 'float'>)
time (s)  5.06827
Numpy sampler (dtype <class 'float'>)
time (s)  10.787635

Reply:
@chrisrothUT did you try with gpu?

Reply:
For n =2 chains

Jax sampler (dtype <class 'float'>)
time (s)  6.84447
Numpy sampler (dtype <class 'float'>)
time (s)  30.603344

Reply:
That was a cpu. I will try on the GPU whenever I manage to install NetKet